app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java

```diff
@@ -117,29 +119,53 @@ private boolean exportTrackPoint(XmlSerializer ser, ActivityPoint point, String
         if (description != null) {
             ser.startTag(NS_DEFAULT, "desc").text(description).endTag(NS_DEFAULT, "desc");
         }
-        ser.startTag(NS_DEFAULT, "src").text(source).endTag(NS_DEFAULT, "src");
+        //ser.startTag(NS_DEFAULT, "src").text(source).endTag(NS_DEFAULT, "src");
 
-        exportTrackpointExtensions(ser, point);
+        exportTrackpointExtensions(ser, point, trackPoints);
 
         ser.endTag(NS_DEFAULT, "trkpt");
 
         return true;
     }
 
-    private void exportTrackpointExtensions(XmlSerializer ser, ActivityPoint point) throws IOException {
+    private void exportTrackpointExtensions(XmlSerializer ser, ActivityPoint point, List<ActivityPoint> trackPoints) throws IOException {
         if (!includeHeartRate) {
             return;
         }
 
         int hr = point.getHeartRate();
         if (!HeartRateUtils.isValidHeartRateValue(hr)) {
+            if(!includeHeartRateByTime) { return; }
+
+            Date time = point.getTime();
+            ActivityPoint closestPointItem = null;
+            long lowestDifference = Long.MAX_VALUE;
+            for (ActivityPoint pointItem : trackPoints) {
+                int hrItem = pointItem.getHeartRate();
+                if(HeartRateUtils.isValidHeartRateValue(hrItem)) {
+                    Date timeItem = pointItem.getTime();
+                    if (timeItem.after(time) || timeItem.equals(time)) continue;
+                    long difference = time.getTime() - timeItem.getTime();
+                    if (difference < lowestDifference) {
+                    lowestDifference = difference;
+                    closestPointItem = pointItem;
+                    }
+                }
+            }
+            if(closestPointItem != null) {
+                hr = closestPointItem.getHeartRate();
+                if (!HeartRateUtils.isValidHeartRateValue(hr)) {
+                    return;
+                }
+            }
             return;
```

Comment:
BTW, this return here made the whole nearest sample finding useless, AFAICS.

Reply:
Quite possible, I did notice that the HR samples were still only appearing every 5th track point.  I didn't get a chance to dig any further.